### PR TITLE
Remove ActiveSupport from runtime

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Relates to https://github.com/Homebrew/brew/issues/16190

Removes ActiveSupport from the brew runtime (it will still exist as a transitive dependency of rubocop-rails, for now).

I recommend to squash commits on merge (or I can force-push a squashed commit, not sure what the merge options are here).

I've used just the parts of the ActiveSupport files that we use, with some light modifications (including adding type signatures, and pulling in upstream changes, for which I've added comments). Another option could be to just vendor the existing dependencies, as we do with `mechanize`. However, ActiveSupport sometimes undergoes refactors and dependency changes, which might be a headache to keep up with (for instance, [this refactor](https://github.com/rails/rails/commit/43b980368a7628fac95cc4f673e0dfbcee77c10b) of Hash#deep_merge! to introduce another `DeepMergeable` module). 

I haven't added new tests for this functionality, but don't want to dismiss their importance/value (rails uses a different framework, preventing this from being a simple c/p operation). Looks like there a just few spots without coverage though (including some error paths that may be hard to plausibly simulate). LMK if more are desirable.

### CODE REVIEW

File: Library/Homebrew/tab.rb [hunk 1]
```diff
@@ -92,7 +92,7 @@ def self.from_file_content(content, path)
     end
 
     if attributes["source"]["spec"].nil?
-      version = PkgVersion.parse path.to_s.split("/").second_to_last
+      version = PkgVersion.parse path.to_s.split("/")[-2]
```
Comment (Bo98) [inline Library/Homebrew/tab.rb:95]: `File.basename(File.dirname(path))` might be a lot clearer what's going on here.

File: Library/Homebrew/download_strategy.rb [hunk 2]
```diff
@@ -360,7 +360,7 @@ def parse_basename(url, search_query: true)
       end
 
       if search_query && (uri_query = uri.query.presence)
-        components[:query] = URI.decode_www_form(uri_query).map(&:second)
+        components[:query] = URI.decode_www_form(uri_query).map { _1.fetch(1) }
```
Comment (Bo98) [inline Library/Homebrew/download_strategy.rb:363]:
```suggestion
        components[:query] = URI.decode_www_form(uri_query).map { |_, value| value }
```

maybe

File: Library/Homebrew/tab.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/tab.rb:95]: Or rely on `Pathname` and use `parent`

File: Library/Homebrew/utils/gems.rb [hunk 3]
```diff
@@ -14,7 +14,7 @@ module Homebrew
 
   # Bump this whenever a committed vendored gem is later added to gitignore.
   # This will trigger it to reinstall properly if `brew install-bundler-gems` needs it.
-  VENDOR_VERSION = 4
+  VENDOR_VERSION = 5
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/gems.rb:17]: @Bo98 Does this also need to be bumped each time a line is removed from `.gitignore` like we're doing? If not: let's not. If so: let's update the comment.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far!

File: Library/Homebrew/utils/gems.rb (hunk 3 above)
Comment (Bo98) [reply Library/Homebrew/utils/gems.rb:17]: With ActiveSupport we gitignore the whole gem and have exceptions for individual files.

So removing an exception from gitignore = adding to git's ignore list if that makes sense

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/gems.rb:17]: Gotcha.

File: Library/Homebrew/utils/gems.rb [hunk 4]
```diff
@@ -14,7 +14,7 @@ module Homebrew
 
   # Bump this whenever a committed vendored gem is later added to gitignore.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/gems.rb:15]:
```suggestion
  # Bump this whenever a committed vendored gem is later added to or exclusion removed from gitignore.
```

File: Library/Homebrew/extend/array.rb [hunk 5]
```diff
@@ -1,7 +1,17 @@
...
+  # Equal to <tt>self[1]</tt>.
+  #
+  #   %w( a b c d e ).second # => "b"
+  def second = self[1]
+
+  # Equal to <tt>self[2]</tt>.
+  #
+  #   %w( a b c d e ).third # => "c"
+  def third = self[2]
+
```
Comment (bevanjkay) [inline Library/Homebrew/extend/array.rb:14]: I'm sure you will see it in the CI errors, but we do use `array.fourth` occasionally in `homebrew-cask` when a version strings contains four parts.

Comment (dduugg) [reply Library/Homebrew/extend/array.rb:14]: I appreciate the heads up nonetheless, ty!

Comment (bevanjkay) [reply Library/Homebrew/extend/array.rb:14]: Looks like we also use fifth! That I did not know.

File: Library/Homebrew/cli/parser.rb [hunk 6]
```diff
@@ -126,9 +126,9 @@ def initialize(&block)
         # Filter out Sorbet runtime type checking method calls.
         cmd_location = T.must(caller_locations).select do |location|
           T.must(location.path).exclude?("/gems/sorbet-runtime-")
-        end.second
-        @command_name = cmd_location.label.chomp("_args").tr("_", "-")
-        @is_dev_cmd = cmd_location.absolute_path.start_with?(Commands::HOMEBREW_DEV_CMD_PATH)
+        end.fetch(1)
+        @command_name = T.must(cmd_location.label).chomp("_args").tr("_", "-")
+        @is_dev_cmd = T.must(cmd_location.absolute_path).start_with?(Commands::HOMEBREW_DEV_CMD_PATH)
```
Comment (dduugg) [inline Library/Homebrew/cli/parser.rb:131]: Some minor changes that are required when adding `sig`s to vendored methods

File: Library/Homebrew/extend/array.rbi [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# typed: strict
+
+class Array
+  sig { returns(T.nilable(Elem)) }
```
Comment (dduugg) [inline Library/Homebrew/extend/array.rbi:4]: `Elem` is an [upstream](https://github.com/sorbet/sorbet/blob/5e26c51/rbi/core/array.rbi#L352) RBI creation that isn't available at runtime, so these need to be in RBI files. (And similarly for `Enumerable::Enum`, `Hash::K`, `Hash::V`, etc. below…)

File: Library/Homebrew/extend/enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  #
+  #   { a: "", b: 1, c: nil, d: [], e: false, f: true }.compact_blank
+  #   # => { b: 1, f: true }
+  sig { returns(T.self_type) }
+  def compact_blank = T.unsafe(self).reject(&:blank?)
+end
+
+class Hash
+  # Hash#reject has its own definition, so this needs one too.
+  def compact_blank = reject { T.unsafe(_2).blank? }
```
Comment (dduugg) [inline Library/Homebrew/extend/enumerable.rb:47]: Happy to write this as `|_k, v| T.unsafe(v).blank?`, i know some folks aren't into numbered block params

File: Library/Homebrew/extend/file/atomic.rb [hunk 9]
```diff
@@ -38,32 +46,35 @@ def self.atomic_write(file_name, temp_dir = dirname(file_name))
...
-      rename(temp_file.path, file_name)
+      rename(T.must(temp_file.path), file_name)
       return_val
     end
   end
 
   # Private utility method.
-  def self.probe_stat_in(dir) #:nodoc:
+  sig { params(dir: String).returns(T.nilable(File::Stat)) }
+  private_class_method def self.probe_stat_in(dir) # :nodoc:
```
Comment (dduugg) [inline Library/Homebrew/extend/file/atomic.rb:65]: visibility modifier added due to code comment above

File: Library/Homebrew/extend/file/atomic.rb [hunk 10]
```diff
@@ -38,32 +46,35 @@ def self.atomic_write(file_name, temp_dir = dirname(file_name))
...
     basename = [
       ".permissions_check",
       Thread.current.object_id,
       Process.pid,
-      rand(1000000)
+      rand(1_000_000),
```
Comment (dduugg) [inline Library/Homebrew/extend/file/atomic.rb:70]: rubocop formatting change

File: Library/Homebrew/extend/hash/deep_merge.rb [hunk 11]
```diff
@@ -22,10 +23,10 @@ def deep_merge(other_hash, &block)
...
   def deep_merge!(other_hash, &block)
     merge!(other_hash) do |key, this_val, other_val|
-      if this_val.is_a?(Hash) && other_val.is_a?(Hash)
-        this_val.deep_merge(other_val, &block)
-      elsif block_given?
-        block.call(key, this_val, other_val)
+      if T.unsafe(this_val).is_a?(Hash) && other_val.is_a?(Hash)
+        T.unsafe(this_val).deep_merge(other_val, &block)
+      elsif block
+        yield(key, this_val, other_val)
```
Comment (dduugg) [inline Library/Homebrew/extend/hash/deep_merge.rb:29]: rubocop-performance wanted these changes ¯\_(ツ)_/¯

File: Library/Homebrew/extend/enumerable.rbi [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  sig {
+    type_parameters(:key).params(
+      block: T.nilable(T.proc.params(o: Enumerable::Elem).returns(T.type_parameter(:key))),
+    ).returns(T::Hash[T.type_parameter(:key), Enumerable::Elem])
+  }
+  def index_by(&block); end
+end
+
+class Hash
+  sig { returns(T::Hash[Hash::K, Hash::V]) }
```
Comment (dduugg) [inline Library/Homebrew/extend/enumerable.rbi:15]: I would like to use a runtime `sig` with `returns(T.self_type)`, but that's a bug: https://github.com/sorbet/sorbet/issues/7586

File: Library/Homebrew/extend/object/deep_dup.rb [hunk 13]
```diff
@@ -40,16 +43,35 @@ class Hash
...
         hash.delete(key)
-        hash[key.deep_dup] = value.deep_dup
+        hash[T.unsafe(key).deep_dup] = T.unsafe(value).deep_dup
       end
     end
     hash
   end
 end
+
+class Module
```
Comment (dduugg) [inline Library/Homebrew/extend/object/deep_dup.rb:62]: This is sourced from the current [upstream](https://github.com/rails/rails/pull/48106/files#diff-1219ce839b402dee52bc46245a0d30d13d961f2089e8cd306aedb978139792db) implementation

File: Library/Homebrew/extend/object/duplicable.rb [hunk 14]
```diff
@@ -23,27 +24,34 @@ class Object
...
-  end
+  #   method(:puts).unbind.duplicable? # => false
+  #   method(:puts).unbind.dup         # => TypeError: allocator undefined for UnboundMethod
+  sig { returns(FalseClass) }
+  def duplicable? = false
+end
+
+require "singleton"
+
+module Singleton
```
Comment (dduugg) [inline Library/Homebrew/extend/object/duplicable.rb:51]: another [upstream](https://github.com/rails/rails/pull/41095/files) addition

File: Library/Homebrew/livecheck/strategy/gnu.rb [hunk 15]
```diff
@@ -32,11 +32,11 @@ class Gnu
         NICE_NAME = "GNU"
 
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = %r{
+        URL_MATCH_REGEX = T.let(%r{
           ^https?://
           (?:(?:[^/]+?\.)*gnu\.org/(?:gnu|software)/(?<project_name>[^/]+)/
           |(?<project_name>[^/]+)\.gnu\.org/?$)
-        }ix.freeze
+        }ix.freeze, Regexp)
```
Comment (dduugg) [inline Library/Homebrew/livecheck/strategy/gnu.rb:39]: Strict typing and this change somehow resolve a spurious type error on line 42 ¯\_(ツ)_/¯

File: Library/Homebrew/standalone/sorbet.rb [hunk 16]
```diff
@@ -2,6 +2,7 @@
 # frozen_string_literal: true
 
 require "sorbet-runtime"
+require "extend/module"
```
Comment (dduugg) [inline Library/Homebrew/standalone/sorbet.rb:5]: This is mostly unrelated, but as of [this](https://github.com/Homebrew/brew/pull/16406/files#diff-eb8b4c210b5cc8f5d6cbbedc40fac8201d598efc5d4310002a9066d0b6ce4902) change, the extension only extends `Module` to `include T::Sig`, so it makes the most sense to live here IMO.

File: Library/Homebrew/tab.rb (hunk 1 above)
Comment (dduugg) [reply Library/Homebrew/tab.rb]: Good call, done

File: Library/Homebrew/download_strategy.rb (hunk 2 above)
Comment (dduugg) [reply Library/Homebrew/download_strategy.rb]: These simplify to `_2`, which seems semantically equivalent to `second`, but i'm open to feedback that it's a bit too arcane.

File: .gitignore [hunk 17]
```diff
@@ -53,25 +53,6 @@
 !**/vendor/bundle/ruby/*/gems/*/lib
 !**/vendor/bundle/ruby/*/gems/addressable-*/data
 !**/vendor/bundle/ruby/*/gems/public_suffix-*/data
-!**/vendor/bundle/ruby/*/gems/rubocop-performance-*/config
-!**/vendor/bundle/ruby/*/gems/rubocop-rails-*/config
-!**/vendor/bundle/ruby/*/gems/rubocop-rspec-*/config
-!**/vendor/bundle/ruby/*/gems/rubocop-sorbet-*/config
```
[the remark this answers is not in the document]
Comment (Bo98) [reply .gitignore:59]: yeah these aren't needed

**Comment by dduugg** [comment]
Ready for review

File: Library/Homebrew/extend/enumerable.rb [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  #
+  #   people.index_by { |person| "#{person.first_name} #{person.last_name}" }
+  #   # => { "Chade- Fowlersburg-e" => <Person ...>, "David Heinemeier Hansson" => <Person ...>, ...}
+  def index_by(&block)
+    if block
+      result = {}
+      each { |elem| result[yield(elem)] = elem }
+      result
+    else
+      T.unsafe(self).to_enum(:index_by) { T.unsafe(self).size if respond_to?(:size) }
```
Comment (apainintheneck) [inline Library/Homebrew/extend/enumerable.rb:19]: This is really ugly.

File: Library/Homebrew/extend/enumerable.rb (hunk 8 above)
Comment (apainintheneck) [reply Library/Homebrew/extend/enumerable.rb:47]: I think that would be clearer in this case since it's a hash.

**Review by apainintheneck — APPROVED** [review]

File: Library/Homebrew/extend/enumerable.rb (hunk 18 above)
Comment (Bo98) [reply Library/Homebrew/extend/enumerable.rb:19]: I'm honestly not entirely sure what's going on here. There's only two usages of `index_by` so if we're wanting to keep it then we don't really need to keep the nil-block case which has a behaviour that's undocumented anyway.

The whole thing can probably be simpified to `each_with_object({}) { |elem, hash| hash[yield(elem)] = elem }`, or replace the two usages with that.

**Comment by Bo98** [comment]
`deep_dup` and friends can probably also go - I'm not fully convinced about its current usage does much. But given this PR is already quite large I'm ok deferring that to a separate PR to reduce risk.

File: Library/Homebrew/extend/enumerable.rb (hunk 18 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/enumerable.rb:19]: I agree that `index_by` can probably be removed entirely and this code inlined given there's only two usages.

File: Library/Homebrew/extend/object/deep_dup.rb [hunk 19]
```diff
@@ -12,6 +13,7 @@ class Object
   #
   #   object.instance_variable_defined?(:@a) # => false
   #   dup.instance_variable_defined?(:@a)    # => true
+  sig { returns(T.self_type) }
   def deep_dup
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/object/deep_dup.rb:17]: As @Bo98 mentioned: I think we can drop this.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @dduugg! As this is a bit of a handful to review, I think it's probably worth merging soon (or as-is, will leave that up-to-you).

I think `index_by` and perhaps `deep_dup` could be dropped but that could be a follow-up PR if necessary.

File: Library/Homebrew/extend/object/deep_dup.rb (hunk 19 above)
Comment (dduugg) [reply Library/Homebrew/extend/object/deep_dup.rb:17]: Naïvely swapping out `deep_dup` for `dup` in `abstract_artifact` leads to test failures in `test/cask/cask_spec.rb` – could i trouble someone with more context to take a deeper (pun intended) look into this one?

**Comment by dduugg** [comment]
I believe that I've addressed all the feedback except for `deep_dup`, PTAL

**Review by MikeMcQuaid — APPROVED** [review]
Thanks again @dduugg!

**Pull request merged**